### PR TITLE
Add an explainer for the extension document

### PR DESCRIPTION
**Pull request by NavidZ**
Add a brief explainer for TAG review for the following features:
 - getCoalescedEvents
 - getPredictedevents
 - pointerrawupdate

### CODE REVIEW

**Review by EiraGe — APPROVED** [review]

File: explainer.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
+# Extension document explainer
+This document is a brief explainer for new features added in the [extension document](https://w3c.github.io/pointerevents/extension.html).
+
+## Coalesced Events
+Nowadays there are high frequency devices that generate user inputs much faster than frame refresh rate in the apps.
```
Comment (patrickhlauke) [inline explainer.md]: "than frame refresh" > "than the frame refresh"

File: explainer.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+Hence some browsers chose to not deliver all those events to the web pages as they realized it is going to be a lot of extra work that doesn't necessary have an effect on the visual output.
```
Comment (patrickhlauke) [inline explainer.md]: "doesn't necessary" > "doesn't necessarily"

File: explainer.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+For example Chromium delays all move events until right before rAF and Firefox does the same if the first event in the frame causes any changes to the DOM.
+
+### Problem
+Since most of the web apps care only about the latest state of the input before generating the next frame, the events happened during the frame could be coalesced together into one event that gets dispatched to the page.
```
Comment (patrickhlauke) [inline explainer.md]: the "could be coalesced" sounds hypothetical. Instead, unless I'm misunderstanding, this is describing what is actually happening currently. So better to just say "are currently coalesced" ?

File: explainer.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+In this case the dispatched event could have the latest screenX/Y coordinates and its movementX/Y could be the sum of movementX/Y of all the events that got coalesced into the dispatch event.
```
Comment (patrickhlauke) [inline explainer.md]: as per previous comment, if this is describing current state, best to drop the "could", so just "the dispatched event has the latest screenX/Y and its movementX/Y is the sum ..."

File: explainer.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+This greatly helped reduce the amount of unnecessary work by UA for routing, hittesting and such for most of the web apps.
```
Comment (patrickhlauke) [inline explainer.md]: "hittesting" > "hit-testing"

File: explainer.md [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+However, there is small set of apps like drawing apps that might still care about the historical details of all those events.
+For instance they need to know the exact path of the pointer to be able to draw a smooth line.
+So although coalescing events and dispatching the coalesced event once per frame would greatly help in terms of performance we still need an API on the dipatched event to expose the coalesced events information in case the app wants to take advantage of that.
```
Comment (patrickhlauke) [inline explainer.md]: per previous comment, if this is what's happening now, change "once per frame would greatly help" to just "once per frame greatly helps"

File: explainer.md [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+### Solution
+We propose to provide an API called `getCoalescedEvents` on `PointerEvent` object so that it exposes any events that have been coalesced into the dispatched event.
```
Comment (patrickhlauke) [inline explainer.md]: "on the `PointerEvent` object that exposes"

File: explainer.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+This API also matches other platforms such [getHistoricalX/Y](https://developer.android.com/reference/android/view/MotionEvent.html#getHistoricalX(int,%20int)) on Android and [similar API](https://developer.apple.com/documentation/uikit/touches_presses_and_gestures/handling_touches_in_your_view/getting_high-fidelity_input_with_coalesced_touches) on iOS.
```
Comment (patrickhlauke) [inline explainer.md]: "such as [getHistoricalX/Y]"

File: explainer.md [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+There are two options for exposing the coalesced events:
+1. Return a list of dictionary of interesting and useful attributes such as cooridnates.
```
Comment (patrickhlauke) [inline explainer.md]: "cooridnates" > "coordinates"

File: explainer.md [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+2. Return a list of `PointerEvent` object representing the actual events that have been coalesced.
```
Comment (patrickhlauke) [inline explainer.md]: "of `PointerEvent` object" > "of `PointerEvent` objects"

File: explainer.md [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+Each item in the result list corresponds to one coalesced event.
+The first option has the benefit of not having an actual `Event` object and hence there is no need to deal with set of functionalities that we don't need them on coaleaced event suchs as propagation functions in the specification.
```
Comment (patrickhlauke) [inline explainer.md]: "no need to deal with a set of functionalities that we don't need on coalesced events, such as..."

File: explainer.md [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+The second option gives a more usable way to the developers since if their code is capable of handling a `PointerEvent` object already they could easily just pass the coalesced events to that and it would work fine.
+
+Because of the ease of use for the developers and also after getting some feedback from the developer, the working group chose the second option.
```
Comment (patrickhlauke) [inline explainer.md]: "Because of the ease of use for developers, and following positive feedback, the working group..."

File: explainer.md [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+You can fine examples on how to use the proposed API in the [proposed specification](https://w3c.github.io/pointerevents/extension.html#examples).
```
Comment (patrickhlauke) [inline explainer.md]: "fine" > "find"

File: explainer.md [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+
+## Predicted Events
+This is the proposal to match other platforms predicted events API for the events to reduce the general user perception of latency.
```
Comment (patrickhlauke) [inline explainer.md]: "other platforms predicted" > "other platforms' predicted"

File: explainer.md [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+
+### Problem
+Imagine a drawing app that draws the points as it gets from an input device such as mouse.
```
Comment (patrickhlauke) [inline explainer.md]: "as it gets from" > "as it gets them from"

File: explainer.md [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+There is always some amount of latency compare to when the users draw on a paper due to inherent UA input pipeline latency and graphics pipeline latency and buffering.
```
Comment (patrickhlauke) [inline explainer.md]: "latency compare to" > "latency compared to"

File: explainer.md [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+So having some sort of the prediction to draw preemptively a little in the future would reduce the perception of the latency. Having the API as part of the web platform enables UA to use some of the underlying platform APIs as well as using other attributes that might not be exposed to the web directly in the prediction algorithm.
+
+### Solution
+We propose to provide an API called `getPredictedEvents` on `PointerEvent` object. This API format will have the same concerns and API alternatives as `getCoalescedEvent` API introduced earlier.
```
Comment (patrickhlauke) [inline explainer.md]: "on `PointerEvent` object" > "on the `PointerEvent` object"

Comment (patrickhlauke) [reply explainer.md]: "as `getCoalescedEvent` API" > "as the `getCoalescedEvent` API"

File: explainer.md [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+You can fine examples on how to use the proposed API in the [proposed specification](https://w3c.github.io/pointerevents/extension.html#examples).
+
+
+## Predicted Events
+This is the proposal to match other platforms predicted events API for the events to reduce the general user perception of latency.
+
+### Problem
+Imagine a drawing app that draws the points as it gets from an input device such as mouse.
+There is always some amount of latency compare to when the users draw on a paper due to inherent UA input pipeline latency and graphics pipeline latency and buffering.
+So having some sort of the prediction to draw preemptively a little in the future would reduce the perception of the latency. Having the API as part of the web platform enables UA to use some of the underlying platform APIs as well as using other attributes that might not be exposed to the web directly in the prediction algorithm.
```
Comment (patrickhlauke) [inline explainer.md]: drop the "So"

File: explainer.md [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+
+### Solution
+We propose to provide an API called `getPredictedEvents` on `PointerEvent` object. This API format will have the same concerns and API alternatives as `getCoalescedEvent` API introduced earlier.
+So in order to match it with `getCoalescedEvents` as well as exposing an ergonomic API we will use the same format as `getCoalescedEvents` by returning a list of `PointerEvent` objects.
```
Comment (patrickhlauke) [inline explainer.md]: drop the "So"

File: explainer.md [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+The events in the result list will have timestamps in the future and UA believes those events are going to be where the pointer is heading.
```
Comment (patrickhlauke) [inline explainer.md]: "in the future and UA believes those events..." > "in the future, and the coordinates will be the position the UA predicts the pointer is heading to."

File: explainer.md [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+This solution will match other platforms such as [prediction](https://developer.apple.com/documentation/uikit/touches_presses_and_gestures/handling_touches_in_your_view/minimizing_latency_with_predicted_touches) on iOS.
+You can fine examples on how to use the proposed API in the [proposed specification](https://w3c.github.io/pointerevents/extension.html#examples).
```
Comment (patrickhlauke) [inline explainer.md]: "fine" > "find"

File: explainer.md [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+
+## PointerRawUpdate
+Nowdays some UAs delay the events until before the next rAF as explained earlier in this document.
+For example Chromium delays all move events until right before rAF and Firefox does the same if the first event in the frame causes any changes to the DOM.
+However, there is another low latency need that this way of delaying events would break.
+
+### Problem
+There are some apps that may need to handle the events regardless of any visual effect.
+For example, some of the apps might be sending events over the network like in the cloud gaming use cases and some others might interact with audio devices.
+Either way delaying the delivery of the events would hurt the experience on these apps. So this problem needs a way of delivering the events without any delay.
```
Comment (patrickhlauke) [inline explainer.md]: drop the "Either way". in the second sentence, I'd change "So this problem needs a way" to "The app needs a way"

File: explainer.md [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+
+### Solution
+Although delivering the events at the speed they happen might have a general burden on UA and running input pipeline for all those events might hurt the performance, this is still necessary for the aformentioned apps.
+So we propose to have a new event type called `pointerrawupdate` to capture this high frequency event.
```
Comment (patrickhlauke) [inline explainer.md]: drop the "So"

File: explainer.md [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+Note that this needs to be a new event to guarantee if the web apps chose to listen to it they understand that UA needs to do extra work for them and may not be able to optimize the input pipeline as much.
```
Comment (patrickhlauke) [inline explainer.md]: "that UA needs" > "that the UA needs"

File: explainer.md [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+This event is supposed to be dispatched as soon as possible without any intentional delaying from UA for any visual optimization such as delaying event to the next rAF.
```
Comment (patrickhlauke) [inline explainer.md]: "from UA" > "from the UA"

**Review by patrickhlauke — CHANGES_REQUESTED** [review]
Apart from the grammatical changes, LGTM

**Comment by NavidZ** [comment]
Thanks Patrick. You made it so easy to apply the comments when you just say what should be there :). Appreciated.

**Review by patrickhlauke — APPROVED** [review]

**Comment by patrickhlauke** [comment]
> Thanks Patrick. You made it so easy to apply the comments when you just say what should be there :). Appreciated.

no worries. next time, i'll work out how to make the edits directly on your PR :)

**Pull request merged**